### PR TITLE
[WIP] Support u32 tokens for llama using Table/InMemoryTable trait objects

### PR DESCRIPTION
**Pull request by luciaquirke**
enum-type not actually enum solution to selecting the correct suffix table token type

### CODE REVIEW

File: src/table.rs [hunk 1]
```diff
@@ -338,6 +334,124 @@ where
...
+            .filter_map(|&item| E::try_from(item).ok())
+            .collect();
+        self.count_next(&query)
+    }
+
+    fn batch_count_next(&self, queries: &[Vec<usize>]) -> Vec<Vec<usize>> {
+        queries
+            .into_par_iter()
+            .map(|query| {
+                let query = unsafe { std::mem::transmute(query.as_slice()) };
```
Comment (norabelrose) [inline src/table.rs:444]: Wait why do we need to do an unsafe transmute here? Wouldn't `as_slice` suffice?

**Review by norabelrose — COMMENTED** [review]
Wondering about the unsafe transmute thing

File: src/table.rs (hunk 1 above)
Comment (luciaquirke) [reply src/table.rs:444]: ahh sorry I thought this was gone, let me get rid of it

File: src/in_memory_index.rs [hunk 2]
```diff
@@ -8,44 +8,103 @@ use std::io::Read;
...
+
+        let table: Box<dyn InMemoryTable + Send + Sync> = if vocab <= u16::MAX as usize + 1 {
+            let tokens: Vec<u16> = tokens.iter().map(|&x| x as u16).collect();
+            Box::new(SuffixTable::<Box<[u16]>, Box<[u64]>>::new(tokens, Some(vocab), verbose))
+        } else {
+            let tokens: Vec<u32> = tokens.iter().map(|&x| x as u32).collect();
+            Box::new(SuffixTable::<Box<[u32]>, Box<[u64]>>::new(tokens, Some(vocab), verbose))
+        };
+
+        assert!(table.is_sorted());
```
[the remark this answers is not in the document]
Comment (luciaquirke) [reply src/in_memory_index.rs]: Done, thanks

**Review by norabelrose — COMMENTED** [review]
One println

**Pull request merged**